### ISSUE TITLE
[3.5.1] [4.0.rc1] Calling GPUParticles3D.restart() from AnimationPlayer doesn't seem to have an effect

### ISSUE THREAD

**Issue by hmans** [godotengine/godot#73043]
### Godot version

3.5.1, 4.0.rc1

### System information

macOS, Apple M1

### Issue description

Resetting a GPU particle effect by invoking its `restart()` function from an animation call track doesn't seem to work.

In the attached MRP, I have a one-shot GPU particle effect that is supposed to play at the start of my animation. For this, I've set up the animation to do two things:

- Set `emitting` to true (this works)
- Call `restart()`, in case the effect is still emitting when the animation starts

But the call to `restart()` doesn't seem to have any effect.

**Notes:**

- I observed this in both 3.5.1 and 4.0.rc1. The attached MRP uses Godot 3.5.1.
- It doesn't matter where on the animation timeline I call `restart()`, it never does anything.
- This makes working on visual effects composed from multiple layered one-short particle effects difficult because now when you hit Shift-D to restart the animation, the effects who were still emitting at that time will not restart.
- I don't know if CPU particles are also affected.

**Screen recording:**

https://user-images.githubusercontent.com/1061/218132825-21641f20-d17c-42c3-b73b-d01b6086e11b.mov

### Steps to reproduce

- Open the attached MRP
- Edit the animation timeline
- Hit Shift-D to start the animation
- Hit it again a few times to restart it while it's still playing

### Minimal reproduction project

[MRP-Godot3-VFX-Timeline.zip](https://github.com/godotengine/godot/files/10708995/MRP-Godot3-VFX-Timeline.zip)

**Comment by hmans** [comment]
Just want to confirm that explicitly clicking on "Restart" in the GPU particles toolbar has the intended effect (even while the effect is still playing), both in 3.5.1 and 4.0.

![image](https://user-images.githubusercontent.com/1061/218136419-bf6019be-5e10-4191-a06f-84525e8c1a7d.png)

**Comment by TokageItLab** [comment]
I remember @reduz saying something earlier about the necessity of a TrackType for particles.

**Comment by hmans** [comment]
A quick update. It seems that what's happening here is that Call Method tracks will only actually _call methods_ when the scene or game is being run, but not within the editor. Which _probably_ makes sense (you wouldn't want to run `queue_destroy()` while editing, etc.).

The documentation unfortunately doesn't mention this (at least I couldn't find anything about it), so if this is by design, I think the documentation should be updated accordingly.

Maybe a toggle can be edited to keyframe events in Method Call tracks so that, when set, the method is also invoked from within the editor?

Either way, the issue remains that it's extremely difficult to work on animations that orchestrate a number of one-shot particle effects. Being able to restart particle effects through animations would improve things significantly.

**Comment by hmans** [comment]
For anyone finding this, here's a quick and dirty workaround:

**Step 1:** create a custom class that extends GPUParticles3D and adds a new virtual property:

```gdscript
@tool

class_name HackedGPUParticles3D
extends GPUParticles3D

@export var hacked_emitting : bool:
  get: return emitting
  set(v):
    emitting = v
    restart()
```

**Step 2:** Use this for your GPU particle effects, or change the types of existing `GPUParticle3D` nodes to it (might require a scene reload/editor restart.)

**Step 3:** Instead of using a Call Method track, use a normal Property track, and set that property to true.

**Comment by clayjohn** [comment]
The fact that methods won't be called when running in the editor should be well documented (and possibly should be communicated to the user in editor). 

At the very least it should be added to the docs here: https://docs.godotengine.org/en/latest/tutorials/animation/animation_track_types.html#call-method-tracks

**Comment by TokageItLab** [comment]
> At the very least it should be added to the docs here

Indeed... I'll keep that in mind as I'm organizing the animation documentation now.

**Comment by hmans** [comment]
I'll leave this for you to decide, but in the interest of reducing noise here, maybe this issue should be rewritten or closed in favor of other (possibly new) issues, to better reflect the _actual_ potential changes:

- Documenting that Call Method tracks don't actually do anything in the editor
- Allowing Call method track keyframes to optionally call their method in the editor by way of a new boolean toggle exposed in their inspector (I may post a proposal for this)
- Particle systems being a bit cumbersome to orchestrate in animation timelines. There are some related issues posted here already:
  - #58778
  - #17903
  - I remember seeing more, but couldn't find them.

**Comment by AndySamoil** [comment]
> ss that extends GPUParticles3D and adds a new virtual property
THIS WORKS, this guy is the goat

**Comment by Nodragem** [comment]
I just spend 1h not understanding why my particles were not emitting, to use restart() and test it in-game works. But there was no way to know that in editor.

I would suggest to:
- write a warning ⚠️ in Editor that says the method is only called in-game
- possible allow the user to play it in Editor anyway, as their own risks (i.e. can they do anything worse than with @ tool ?)
